kubernetes/samples/scripts/domain-lifecycle/helper.sh

```diff
@@ -0,0 +1,606 @@
+      done
+    fi
+    staticClause=".domain.configuredClusters[] | select (.dynamicServersConfig == null)"
+    nameCmd=" . | {name: .name, serverName: .servers[].name}"
+    configuredClusters=($(echo $jsonTopology | jq "${staticClause}" | jq -cr "${nameCmd}"))
+    configuredClusterSize=${#configuredClusters[@]}
+    if [ "${configuredClusterSize}" -gt 0 ]; then
+      for configuredClusterName in "${configuredClusters[@]}"; do
+        name=$(echo ${configuredClusterName} | jq -r .serverName)
+        if [ "${serverName}" == "${name}" ]; then
+          eval $__clusterName="'$(echo ${configuredClusterName} | jq -r .name)'"
+          eval $__isValidServer=true
+          break
+        fi
+      done
+    fi
+  fi
+}
+
+#
+# Function to validate whether a cluster is valid and part of the domain
+# $1 - Domain unique id.
+# $2 - Domain namespace.
+# $3 - cluster name
+# $4 - Retrun value indicating whether cluster name is valid
+#
+function validateClusterName {
+  local domainUid=$1
+  local domainNamespace=$2
+  local clusterName=$3
+  local __isValidCluster=$4
+
+  configMap=$(${kubernetesCli} get cm ${domainUid}-weblogic-domain-introspect-cm \
+    -n ${domainNamespace} -o json)
+  topology=$(echo "${configMap}" | jq '.data["topology.yaml"]')
+  jsonTopology=$(python -c \
+    'import sys, yaml, json; print json.dumps(yaml.safe_load('"${topology}"'), indent=4)')
+  clusters=($(echo $jsonTopology | jq -cr .domain.configuredClusters[].name))
+  if  checkStringInArray "${clusterName}" "${clusters[@]}" ; then
+    eval $__isValidCluster=true
+  fi
+}
+
+#
+# check if string passed as first argument is present in array passed as second argument
+# $1 - string to check
+# $2 - array
+checkStringInArray() {
+    local str=$1 arr
+    shift
+    for arr; do
+      [[ $str = "$arr" ]] && return
+    done
+    return 1
+}
+
+# try to execute jq to see whether jq is available
+function validateJqAvailable {
+  if ! [ -x "$(command -v jq)" ]; then
+    validationError "jq is not installed"
+  fi
+}
+
+# try to execute python to see whether python is available
+function validatePythonAvailable {
+  if ! [ -x "$(command -v python)" ]; then
+    validationError "python is not installed"
+  fi
+}
+
+# try to execute kubernetes cli to see whether cli is available
+function validateKubernetesCliAvailable {
+  if ! [ -x "$(command -v ${kubernetesCli})" ]; then
+    validationError "${kubernetesCli} is not installed"
+  fi
+}
+
+#
+# Function to exit and print an error message
+# $1 - text of message
+function fail {
+  printError $*
+  exit 1
+}
+
+# Function to print an error message
+function printError {
+  echo [`timestamp`][ERROR] $*
+}
+
+# Function to print an error message
+function printInfo {
+  echo [`timestamp`][INFO] $*
+}
+
+# timestamp
+#   purpose:  echo timestamp in the form yyyymmddThh:mm:ss.mmm ZZZ
+#   example:  20181001T14:00:00.001 UTC
+function timestamp() {
+  local timestamp="`date --utc '+%Y-%m-%dT%H:%M:%S %N %s %Z' 2>&1`"
```

Comment:
Unfortunately, this fn won't work on the mac bash because you have a global 'set -e'.  To work-around it do this, do a local 'set +e' around the potential illegal call: 
```
  timestamp="$(set +e && date --utc '+%Y-%m-%dT%H:%M:%S %N %s %Z' 2>&1 || exit 0)"
```

Reply:
Note that the 'exit 0' exits from the () scope -- it doesn't actually exit from the entire script...

Reply:
Fixed.